src/lava/lib/optimization/utils/solver_tuner.py

```diff
@@ -21,56 +23,89 @@ def __init__(self, params_grid: ty.Dict):
-
-        for params in params_grid:
-            hyperparameters = dict(zip(params_names, params))
-            solver_parameters["hyperparameters"] = hyperparameters
-            solution = solver.solve(**solver_parameters)
-            # TODO : Implement logic for CSP problems
+             solver_params: ty.Dict,
+             fitness_fn: ty.Callable[[float, int], float],
+             fitness_target: float = None,
+             seed: int = 0):
+        """
+        Perform random search to optimize solver hyper-parameters based on a
+        fitness function.
+
+        Parameters
+        ----------
+        solver: OptimizationSolver
+            Optimization solver to use for solving the problem.
+        solver_params: ty.Dict
+            Parameters for the solver.
+        fitness_fn: ty.Callable[[float, int], float]
+            Fitness function to evaluate a given set of hyper-parameters,
+            taking as input the current cost and number of steps to solution.
+            This is the function that is maximized by the SolverTuner.
+        fitness_target: float, optional
+            Fitness target to reach. If this is not passed, the full grid is
+            explored before stopping search.
+        seed: int, default=0
+            Seed for randomized grid search.
+
+        Returns
+        -------
+        best_hyperparams: ty.Dict
+            Dictionary containing the hyper-parameters with the highest fitness.
+        success: bool
+            Flag signaling if the fitness_target has been reached. If no
+            fitness_target is passed, the flag is True.
+        """
+        # TODO : Check that hyperparams are arguments for solver
+
+        best_hyperparams = None
+        best_fitness = -float('inf')
+
+        random.Random(seed).shuffle(self._params_grid)
+
+        for params in self._params_grid:
+            np.random.seed(seed)
+            hyperparams = dict(zip(self._params_keys, params))
+            solver_params["hyperparameters"] = hyperparams
+            solver.solve(**solver_params)
             cost = solver.last_run_report["cost"]
             step_to_sol = solver.last_run_report["steps_to_solution"]
-            if cost is not None and cost <= best_cost and (
-                    step_to_sol < best_step_to_sol):
-                best_hyperparameters = hyperparameters
-                best_cost = cost
-                best_step_to_sol = step_to_sol
-                print(f"""Better hyperparameters configuration found!\
-                        \nHyperparameters: {best_hyperparameters}\
-                        \nBest cost: {best_cost}\
-                        \nBest step-to-solution: {best_step_to_sol}""")
-            if stopping_condition is not None and (
-                    stopping_condition(best_cost, best_step_to_sol)):
+            fitness = fitness_fn(cost, step_to_sol)
+            self._store_trial(hyperparams, cost, step_to_sol, fitness)
+            if fitness > best_fitness:

```

Comment:
Minor point: I would assume that we want to minimize a function. Thus, the fitness_fn should be defined such that it is minimum for the best set of hyperparameters.

But please feel free to disagree!

Reply:
My point in adopting the fitness (and thus maximization) nomenclature for the SolverTuner is to distinguish it from the cost minimization performed by the Solver. If we find it makes more sense to talk about minimization, though, we can change the name of the argument to something else and change the behavior accordingly. Any suggestions on this?